Support for HTTP GET map parameters?

Do we support passing map parameters using HTTP GET in Armeria?

My code:
```
@Get("")
@ProducesJson
public TablesPagedVO listTables(
		@Param("properties") Optional<Map<String, String>> properties,
		@Param("max_results") Optional<Integer> maxResults,
		@Param("page_number") Optional<Integer> pageNumber) {
	TablesPagedDTO tablesListDTO =
			tableService.listTables(
					properties,
					maxResults,
					pageNumber);
	return VOConverters.toVO(tablesListDTO);
```

An exception occurred:
```
[2025-01-08 15:26:04.411] [armeria-common-worker-nio-2-2] [ERROR] GlobalExceptionHandler.handleException - 
java.lang.IllegalArgumentException: Can't convert '{is_online=false}' to type 'Map'.
	at com.linecorp.armeria.internal.server.annotation.AnnotatedServiceTypeUtil.stringToType(AnnotatedServiceTypeUtil.java:189)
	at com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.convert(AnnotatedValueResolver.java:1103)
	at com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.convert(AnnotatedValueResolver.java:1121)
	at com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.lambda$resolver$22(AnnotatedValueResolver.java:785)
	at com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.resolve(AnnotatedValueResolver.java:1095)
	at com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.lambda$toArguments$1(AnnotatedValueResolver.java:167)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:575)
	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:616)
	at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:622)
	at com.linecorp.armeria.internal.server.annotation.AnnotatedValueResolver.toArguments(AnnotatedValueResolver.java:167)
	at com.linecorp.armeria.internal.server.annotation.DefaultAnnotatedService.invoke(DefaultAnnotatedService.java:405)
	at com.linecorp.armeria.internal.server.annotation.DefaultAnnotatedService.serve0(DefaultAnnotatedService.java:336)
	at com.linecorp.armeria.internal.server.annotation.DefaultAnnotatedService.serve(DefaultAnnotatedService.java:308)
	at com.linecorp.armeria.internal.server.annotation.DefaultAnnotatedService.serve(DefaultAnnotatedService.java:83)
	at com.linecorp.armeria.internal.server.annotation.DefaultAnnotatedService$ExceptionHandlingHttpService.serve(DefaultAnnotatedService.java:584)
	at com.linecorp.armeria.internal.server.RouteDecoratingService.serve(RouteDecoratingService.java:132)
	at com.linecorp.armeria.internal.server.RouteDecoratingService.serve(RouteDecoratingService.java:79)
	at com.linecorp.armeria.server.metric.MetricCollectingService.serve(MetricCollectingService.java:109)
	at com.linecorp.armeria.internal.server.RouteDecoratingService.serve(RouteDecoratingService.java:132)
	at com.linecorp.armeria.internal.server.RouteDecoratingService.serve(RouteDecoratingService.java:79)
	at com.linecorp.armeria.server.logging.LoggingService.serve(LoggingService.java:86)
	at com.linecorp.armeria.internal.server.RouteDecoratingService.serve(RouteDecoratingService.java:132)
	at com.rickyma.uc.server.decorator.AuthenticationDecorator.serve(AuthenticationDecorator.java:54)
	at com.linecorp.armeria.server.FunctionalDecoratingHttpService.serve(FunctionalDecoratingHttpService.java:45)
	at com.linecorp.armeria.internal.server.RouteDecoratingService.serve(RouteDecoratingService.java:132)
	at com.rickyma.uc.server.decorator.ThreadLocalContextDecorator.serve(ThreadLocalContextDecorator.java:18)
	at com.linecorp.armeria.server.FunctionalDecoratingHttpService.serve(FunctionalDecoratingHttpService.java:45)
	at com.linecorp.armeria.internal.server.RouteDecoratingService.serve(RouteDecoratingService.java:132)
	at com.rickyma.uc.server.exception.ExceptionHandlingDecorator.serve(ExceptionHandlingDecorator.java:33)
	at com.linecorp.armeria.server.FunctionalDecoratingHttpService.serve(FunctionalDecoratingHttpService.java:45)
	at com.linecorp.armeria.internal.server.RouteDecoratingService$InitialDispatcherService.serve(RouteDecoratingService.java:202)
	at com.linecorp.armeria.server.HttpServerHandler.serve0(HttpServerHandler.java:490)
	at com.linecorp.armeria.server.HttpServerHandler.handleRequest(HttpServerHandler.java:402)
	at com.linecorp.armeria.server.HttpServerHandler.channelRead(HttpServerHandler.java:282)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at com.linecorp.armeria.server.Http1RequestDecoder.channelRead(Http1RequestDecoder.java:284)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at com.linecorp.armeria.server.HttpServerUpgradeHandler.channelRead(HttpServerUpgradeHandler.java:227)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318)
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:280)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.handler.flush.FlushConsolidationHandler.channelRead(FlushConsolidationHandler.java:152)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1357)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:868)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562)
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

I tried to use a converter, but it seemed only worked when using POST methods:
```
package com.rickyma.uc.server.config;

import com.linecorp.armeria.common.AggregatedHttpRequest;
import com.linecorp.armeria.common.annotation.Nullable;
import com.linecorp.armeria.server.ServiceRequestContext;
import com.linecorp.armeria.server.annotation.RequestConverterFunction;
import java.lang.reflect.ParameterizedType;
import java.util.HashMap;
import java.util.Map;

public class MapRequestConverter implements RequestConverterFunction {

    @Override
    public @Nullable Object convertRequest(
            ServiceRequestContext ctx,
            AggregatedHttpRequest request,
            Class<?> expectedResultType,
            @Nullable ParameterizedType expectedParameterizedResultType)
            throws Exception {
        if (Map.class.isAssignableFrom(expectedResultType)) {
            // Parse the query parameters into a Map
            Map<String, String> result = new HashMap<>();
            String query = request.uri().getRawQuery();
            if (query != null) {
                for (String param : query.split("&")) {
                    String[] keyValue = param.split("=");
                    if (keyValue.length == 2) {
                        result.put(keyValue[0], keyValue[1]);
                    }
                }
            }
            return result;
        }
        return RequestConverterFunction.fallthrough();
    }
}

```

--- a/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
@@ -56,7 +56,6 @@ import java.util.Set;
 import java.util.function.BiFunction;
 import java.util.function.Function;
 import java.util.function.Supplier;
-import java.util.stream.Collectors;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -466,9 +465,7 @@ final class AnnotatedValueResolver {
                 if (DefaultValues.isSpecified(param.value())) {
                     throw new IllegalArgumentException(
                         String.format("Invalid @Param annotation on Map parameter: '%s'. " +
-                                "The @Param annotation specifies a value ('%s'), which is not allowed. " +
-                                "When using a Map, @Param must not specify a value. " +
-                                "Please use @Param Map<String, Object> param without a value.",
+                                "The @Param annotation specifies a value ('%s'), which is not allowed. ",
                             annotatedElement, param.value()));
                 }
                 return ofQueryParamMap(name, annotatedElement, typeElement, type, description);
@@ -611,7 +608,7 @@ final class AnnotatedValueResolver {
             .description(description)
             .aggregation(AggregationStrategy.FOR_FORM_DATA)
             .resolver((resolver, ctx) -> ctx.queryParams().stream()
-                .collect(Collectors.toMap(
+                .collect(toImmutableMap(
                     Entry::getKey,
                     Entry::getValue,
                     (existing, replacement) -> replacement

--- a/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
@@ -465,7 +465,8 @@ final class AnnotatedValueResolver {
                 if (DefaultValues.isSpecified(param.value())) {
                     throw new IllegalArgumentException(
                             String.format("Invalid @Param annotation on Map parameter: '%s'. " +
-                                          "The @Param annotation specifies a value ('%s'), which is not allowed. ",
+                                          "The @Param annotation specifies a value ('%s'), " +
+                                          "which is not allowed. ",
                                           annotatedElement, param.value()));
                 }
                 return ofQueryParamMap(name, annotatedElement, typeElement, type, description);

--- a/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
@@ -48,6 +48,7 @@ import java.util.EnumSet;
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.Objects;
 import java.util.Optional;
 import java.util.ServiceLoader;
@@ -56,6 +57,7 @@ import java.util.function.BiFunction;
 import java.util.function.Function;
 import java.util.function.Supplier;
 
+import java.util.stream.Collectors;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -458,6 +460,11 @@ final class AnnotatedValueResolver {
         final Param param = annotatedElement.getAnnotation(Param.class);
         if (param != null) {
             final String name = findName(typeElement, param.value());
+            // If the parameter is of type Map and the @Param annotation does not specify a value,
+            // map all query parameters into the Map.
+            if (Map.class.isAssignableFrom(type) && DefaultValues.isUnspecified(param.value())) {
+                return ofQueryParamMap(name, annotatedElement, typeElement, type, description);
+            }
             if (type == File.class || type == Path.class || type == MultipartFile.class) {
                 return ofFileParam(name, annotatedElement, typeElement, type, description);
             }
@@ -583,6 +590,25 @@ final class AnnotatedValueResolver {
                                    () -> "Cannot resolve a value from a query parameter: " + name,
                                    queryDelimiter))
                 .build();
+    }
+
+    private static AnnotatedValueResolver ofQueryParamMap(String name,
+        AnnotatedElement annotatedElement,
+        AnnotatedElement typeElement, Class<?> type,
+        DescriptionInfo description) {
+
+        return new Builder(annotatedElement, type, name)
+            .annotationType(Param.class)
+            .typeElement(typeElement)
+            .description(description)
+            .aggregation(AggregationStrategy.FOR_FORM_DATA)
+            .resolver((resolver, ctx) -> ctx.queryParams().stream()
+                .collect(Collectors.toMap(
+                    Entry::getKey,
+                    Entry::getValue,
+                    (existing, replacement) -> replacement
+                )))
+            .build();
     }
 
     private static AnnotatedValueResolver ofFileParam(String name,

--- a/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
@@ -56,8 +56,8 @@ import java.util.Set;
 import java.util.function.BiFunction;
 import java.util.function.Function;
 import java.util.function.Supplier;
-
 import java.util.stream.Collectors;
+
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 

--- a/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
@@ -464,9 +464,9 @@ final class AnnotatedValueResolver {
             if (Map.class.isAssignableFrom(type)) {
                 if (DefaultValues.isSpecified(param.value())) {
                     throw new IllegalArgumentException(
-                        String.format("Invalid @Param annotation on Map parameter: '%s'. " +
-                                "The @Param annotation specifies a value ('%s'), which is not allowed. ",
-                            annotatedElement, param.value()));
+                            String.format("Invalid @Param annotation on Map parameter: '%s'. " +
+                                          "The @Param annotation specifies a value ('%s'), which is not allowed. ",
+                                          annotatedElement, param.value()));
                 }
                 return ofQueryParamMap(name, annotatedElement, typeElement, type, description);
             }
@@ -598,22 +598,22 @@ final class AnnotatedValueResolver {
     }
 
     private static AnnotatedValueResolver ofQueryParamMap(String name,
-        AnnotatedElement annotatedElement,
-        AnnotatedElement typeElement, Class<?> type,
-        DescriptionInfo description) {
+                                                          AnnotatedElement annotatedElement,
+                                                          AnnotatedElement typeElement, Class<?> type,
+                                                          DescriptionInfo description) {
 
         return new Builder(annotatedElement, type, name)
-            .annotationType(Param.class)
-            .typeElement(typeElement)
-            .description(description)
-            .aggregation(AggregationStrategy.FOR_FORM_DATA)
-            .resolver((resolver, ctx) -> ctx.queryParams().stream()
-                .collect(toImmutableMap(
-                    Entry::getKey,
-                    Entry::getValue,
-                    (existing, replacement) -> replacement
-                )))
-            .build();
+                .annotationType(Param.class)
+                .typeElement(typeElement)
+                .description(description)
+                .aggregation(AggregationStrategy.FOR_FORM_DATA)
+                .resolver((resolver, ctx) -> ctx.queryParams().stream()
+                                                .collect(toImmutableMap(
+                                                        Entry::getKey,
+                                                        Entry::getValue,
+                                                        (existing, replacement) -> replacement
+                                                )))
+                .build();
     }
 
     private static AnnotatedValueResolver ofFileParam(String name,

--- a/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
+++ b/core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolver.java
@@ -462,7 +462,15 @@ final class AnnotatedValueResolver {
             final String name = findName(typeElement, param.value());
             // If the parameter is of type Map and the @Param annotation does not specify a value,
             // map all query parameters into the Map.
-            if (Map.class.isAssignableFrom(type) && DefaultValues.isUnspecified(param.value())) {
+            if (Map.class.isAssignableFrom(type)) {
+                if (DefaultValues.isSpecified(param.value())) {
+                    throw new IllegalArgumentException(
+                        String.format("Invalid @Param annotation on Map parameter: '%s'. " +
+                                "The @Param annotation specifies a value ('%s'), which is not allowed. " +
+                                "When using a Map, @Param must not specify a value. " +
+                                "Please use @Param Map<String, Object> param without a value.",
+                            annotatedElement, param.value()));
+                }
                 return ofQueryParamMap(name, annotatedElement, typeElement, type, description);
             }
             if (type == File.class || type == Path.class || type == MultipartFile.class) {

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
@@ -636,8 +636,8 @@ class AnnotatedServiceTest {
         public String map(RequestContext ctx, @Param Map<String, Object> map) {
             validateContext(ctx);
             return map.isEmpty() ? "empty" : map.entrySet().stream()
-                .map(entry -> entry.getKey() + '=' + entry.getValue())
-                .collect(Collectors.joining(", "));
+                                                .map(entry -> entry.getKey() + '=' + entry.getValue())
+                                                .collect(Collectors.joining(", "));
         }
     }
 
@@ -1078,7 +1078,7 @@ class AnnotatedServiceTest {
 
             // Case all query parameters test map
             testBody(hc, get("/7/param/map?key1=value1&key2=value2"),
-                "key1=value1, key2=value2");
+                     "key1=value1, key2=value2");
             testBody(hc, get("/7/param/map"), "empty");
         }
     }
@@ -1379,13 +1379,13 @@ class AnnotatedServiceTest {
     @Test
     void testInvalidParamAnnotationUsageOnMap() {
         assertThatThrownBy(() ->
-            Server.builder()
-                .annotatedService()
-                .build(new MyAnnotationService16())
-                .build()
+                                   Server.builder()
+                                         .annotatedService()
+                                         .build(new MyAnnotationService16())
+                                         .build()
         )
-        .isInstanceOf(IllegalArgumentException.class)
-        .hasMessageContaining("Invalid @Param annotation on Map parameter");
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("Invalid @Param annotation on Map parameter");
     }
 
     private enum UserLevel {

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
@@ -631,7 +631,7 @@ class AnnotatedServiceTest {
         }
 
         @Get("/param/map")
-        public String map (RequestContext ctx, @Param Map<String, Object> map) {
+        public String map(RequestContext ctx, @Param Map<String, Object> map) {
             validateContext(ctx);
             return map.isEmpty() ? "empty" : map.entrySet().stream()
                 .map(entry -> entry.getKey() + "=" + entry.getValue())

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
@@ -26,6 +26,7 @@ import java.nio.charset.StandardCharsets;
 import java.util.ArrayList;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.Map;
 import java.util.Optional;
 import java.util.Set;
 import java.util.TreeSet;
@@ -628,6 +629,14 @@ class AnnotatedServiceTest {
             validateContext(ctx);
             return username + '/' + password;
         }
+
+        @Get("/param/map")
+        public String map (RequestContext ctx, @Param Map<String, Object> map) {
+            validateContext(ctx);
+            return map.isEmpty() ? "empty" : map.entrySet().stream()
+                .map(entry -> entry.getKey() + "=" + entry.getValue())
+                .collect(Collectors.joining(", "));
+        }
     }
 
     @ResponseConverter(UnformattedStringConverterFunction.class)
@@ -1057,6 +1066,11 @@ class AnnotatedServiceTest {
             testStatusCode(hc, get("/7/param/default2"), 400);
 
             testBody(hc, get("/7/param/default_null"), "(null)");
+
+            // Case all query parameters test map
+            testBody(hc, get("/7/param/map?key1=value1&key2=value2"),
+                "key1=value1, key2=value2");
+            testBody(hc, get("/7/param/map"), "empty");
         }
     }
 

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedServiceTest.java
@@ -19,6 +19,7 @@ import static org.apache.hc.core5.http.HttpHeaders.ACCEPT;
 import static org.apache.hc.core5.http.HttpHeaders.CONTENT_TYPE;
 import static org.apache.hc.core5.http.HttpHeaders.IF_MATCH;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 import java.io.IOException;
 import java.nio.charset.Charset;
@@ -76,6 +77,7 @@ import com.linecorp.armeria.common.util.UnmodifiableFuture;
 import com.linecorp.armeria.internal.testing.AnticipatedException;
 import com.linecorp.armeria.internal.testing.GenerateNativeImageTrace;
 import com.linecorp.armeria.server.HttpStatusException;
+import com.linecorp.armeria.server.Server;
 import com.linecorp.armeria.server.ServerBuilder;
 import com.linecorp.armeria.server.ServiceRequestContext;
 import com.linecorp.armeria.server.TestConverters.NaiveIntConverterFunction;
@@ -634,7 +636,7 @@ class AnnotatedServiceTest {
         public String map(RequestContext ctx, @Param Map<String, Object> map) {
             validateContext(ctx);
             return map.isEmpty() ? "empty" : map.entrySet().stream()
-                .map(entry -> entry.getKey() + "=" + entry.getValue())
+                .map(entry -> entry.getKey() + '=' + entry.getValue())
                 .collect(Collectors.joining(", "));
         }
     }
@@ -931,6 +933,13 @@ class AnnotatedServiceTest {
         public ResponseEntity<HttpResponse> responseEntityHttpResponse(RequestContext ctx) {
             validateContext(ctx);
             return ResponseEntity.of(HttpStatus.OK, HttpResponse.of(HttpStatus.UNAUTHORIZED));
+        }
+    }
+
+    public static class MyAnnotationService16 {
+        @Get("/param/map-invalid")
+        public String invalidMapParam(@Param("param") Map<String, String> param) {
+            return "Should not reach here";
         }
     }
 
@@ -1365,6 +1374,18 @@ class AnnotatedServiceTest {
             testStatusCode(hc, get("/17/response-entity-status"), 301);
             testStatusCode(hc, get("/17/response-entity-http-response"), 401);
         }
+    }
+
+    @Test
+    void testInvalidParamAnnotationUsageOnMap() {
+        assertThatThrownBy(() ->
+            Server.builder()
+                .annotatedService()
+                .build(new MyAnnotationService16())
+                .build()
+        )
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("Invalid @Param annotation on Map parameter");
     }
 
     private enum UserLevel {

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
@@ -380,8 +380,8 @@ class AnnotatedValueResolverTest {
                     assertThat(value).isNotNull();
                     assertThat(value).isInstanceOf(Map.class);
                     assertThat((Map<?, ?>) value).size()
-                                                 .isEqualTo(existingHttpParameters.size()
-                                                            + existingWithoutValueParameters.size());
+                                                 .isEqualTo(existingHttpParameters.size() +
+                                                            existingWithoutValueParameters.size());
                 } else {
                     assertThat(value).isEqualTo(resolver.defaultValue());
                 }

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
@@ -380,7 +380,8 @@ class AnnotatedValueResolverTest {
                     assertThat(value).isNotNull();
                     assertThat(value).isInstanceOf(Map.class);
                     assertThat((Map<?, ?>) value).size()
-                        .isEqualTo(existingHttpParameters.size() + existingWithoutValueParameters.size());
+                                                 .isEqualTo(existingHttpParameters.size()
+                                                            + existingWithoutValueParameters.size());
                 } else {
                     assertThat(value).isEqualTo(resolver.defaultValue());
                 }

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
@@ -359,8 +359,6 @@ class AnnotatedValueResolverTest {
                 } else {
                     if (String.class.isAssignableFrom(resolver.elementType())) {
                         assertThat(value).isEqualTo("");
-                    } else if (QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
-                        assertThat(value).isNotNull();
                     } else {
                         assertThat(value).isNull();
                     }

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
@@ -112,6 +112,7 @@ class AnnotatedValueResolverTest {
     static final ServiceRequestContext context;
     static final HttpRequest request;
     static final RequestHeaders originalHeaders;
+    static final String QUERY_PARAM_MAP = "queryParamMap";
     static Map<String, AttributeKey<?>> successExpectAttrKeys;
     static Map<String, AttributeKey<?>> failExpectAttrKeys;
 
@@ -358,12 +359,18 @@ class AnnotatedValueResolverTest {
                 } else {
                     if (String.class.isAssignableFrom(resolver.elementType())) {
                         assertThat(value).isEqualTo("");
+                    } else if(QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
+                        assertThat(value).isNotNull();
                     } else {
                         assertThat(value).isNull();
                     }
                 }
             } else {
-                assertThat(resolver.defaultValue()).isNotNull();
+                if(QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
+                    assertThat(resolver.defaultValue()).isNull();
+                } else {
+                    assertThat(resolver.defaultValue()).isNotNull();
+                }
                 if (resolver.hasContainer() && List.class.isAssignableFrom(resolver.containerType())) {
                     assertThat((List<Object>) value).hasSize(1)
                                                     .containsOnly(resolver.defaultValue());
@@ -371,6 +378,11 @@ class AnnotatedValueResolverTest {
                             .isEqualTo(resolver.elementType());
                 } else if (resolver.shouldWrapValueAsOptional()) {
                     assertThat(value).isEqualTo(Optional.of(resolver.defaultValue()));
+                } else if(QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
+                    assertThat(value).isNotNull();
+                    assertThat(value).isInstanceOf(Map.class);
+                    assertThat((Map<?, ?>) value).size()
+                        .isEqualTo(existingHttpParameters.size() + existingWithoutValueParameters.size());
                 } else {
                     assertThat(value).isEqualTo(resolver.defaultValue());
                 }
@@ -447,6 +459,7 @@ class AnnotatedValueResolverTest {
                      @Param @Default Integer emptyParam2,
                      @Param @Default List<String> emptyParam3,
                      @Param @Default List<Integer> emptyParam4,
+                     @Param Map<String, Object> queryParamMap,
                      @Header List<String> header1,
                      @Header("header1") Optional<List<ValueEnum>> optionalHeader1,
                      @Header String header2,

--- a/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
+++ b/core/src/test/java/com/linecorp/armeria/internal/server/annotation/AnnotatedValueResolverTest.java
@@ -359,14 +359,14 @@ class AnnotatedValueResolverTest {
                 } else {
                     if (String.class.isAssignableFrom(resolver.elementType())) {
                         assertThat(value).isEqualTo("");
-                    } else if(QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
+                    } else if (QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
                         assertThat(value).isNotNull();
                     } else {
                         assertThat(value).isNull();
                     }
                 }
             } else {
-                if(QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
+                if (QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
                     assertThat(resolver.defaultValue()).isNull();
                 } else {
                     assertThat(resolver.defaultValue()).isNotNull();
@@ -378,7 +378,7 @@ class AnnotatedValueResolverTest {
                             .isEqualTo(resolver.elementType());
                 } else if (resolver.shouldWrapValueAsOptional()) {
                     assertThat(value).isEqualTo(Optional.of(resolver.defaultValue()));
-                } else if(QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
+                } else if (QUERY_PARAM_MAP.equals(resolver.httpElementName())) {
                     assertThat(value).isNotNull();
                     assertThat(value).isInstanceOf(Map.class);
                     assertThat((Map<?, ?>) value).size()

